## <center><b>LAB 4 - Multivariate Linear and Polynomial Regression, and Evaluation using R-Squared, MAPE and MAE.</b></center>
<center>Done by: Aardran Premakumar <br> 8844491</center>

In [10]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics as metrics
from sklearn.preprocessing import PolynomialFeatures

In [7]:
#loading data set
df = load_diabetes(as_frame = True, scaled = True)
X = df.data
y = df.target

In [9]:
#splitting the data into training, testing, and valuation sets
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=10)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=10)

## Multivariate Linear Regression on all variables

In [12]:
#training the model
linearModel = LinearRegression()
linearModel.fit(X_train, y_train)

LinearRegression()

In [14]:
#calculating values
yPredLinear = linearModel.predict(X_val)
rSquareLinear = metrics.r2_score(y_val, yPredLinear)
maeLinear = metrics.mean_absolute_error(y_val, yPredLinear)
mapeLinear = metrics.mean_absolute_percentage_error(y_val, yPredLinear)

## Polynomial Regression of the 2nd degree on the BMI feature alone

In [15]:
#transforming the values, setting the degree into 2
polyBMI = PolynomialFeatures(degree=2)
X_train_poly = polyBMI.fit_transform(X_train[['bmi']])
X_val_poly = polyBMI.transform(X_val[['bmi']])


In [16]:
#training the model
polyModelBMI = LinearRegression()
polyModelBMI.fit(X_train_poly, y_train)


LinearRegression()

In [17]:
#calculating values
yPredPolyBMI = polyModelBMI.predict(X_val_poly)
rSquarePolyBMI = metrics.r2_score(y_val, yPredPolyBMI)
maePolyBMI = metrics.mean_absolute_error(y_val, yPredPolyBMI)
mapePolyBMI = metrics.mean_absolute_percentage_error(y_val, yPredPolyBMI)

## Multivariate Polynomial Regression of the 2nd degree on all variables

In [18]:
#transforming the values, setting the degree into 2
polyAll = PolynomialFeatures(degree=2, include_bias=False)
X_train_polyAll = polyAll.fit_transform(X_train)
X_val_polyAll = polyAll.transform(X_val)

In [25]:
#training the model
polyModelAll = LinearRegression()
polyModelAll.fit(X_train_polyAll, y_train)

In [20]:
#calculating values
yPredPolyAll = polyModelAll.predict(X_val_polyAll)

rSquarePolyAll = metrics.r2_score(y_val, yPredPolyAll)
maePolyAll = metrics.mean_absolute_error(y_val, yPredPolyAll)
mapePolyAll = metrics.mean_absolute_percentage_error(y_val, yPredPolyAll)


## Printing all the Values

In [21]:
print("Linear Regression:")
print("R-Squared:",rSquareLinear)
print("Mean Absolute Error:",maeLinear)
print("Mean Absolute Percentage Error:",mapeLinear)

print("\n Polynomial Regression BMI only:")
print("R-Squared:",rSquarePolyBMI)
print("Mean Absolute Error:",maePolyBMI)
print("Mean Absolute Percentage Error:",mapePolyBMI)

print("\n Polynomial Regression on all variables:")
print("R-Squared:",rSquarePolyAll)
print("Mean Absolute Error:",maePolyAll)
print("Mean Absolute Percentage Error:",mapePolyAll)

Linear Regression:
R-Squared: 0.6058418311531759
Mean Absolute Error: 40.29506763793953
Mean Absolute Percentage Error: 0.30227596813542307

 Polynomial Regression BMI only:
R-Squared: 0.41363793815989636
Mean Absolute Error: 51.51886898728833
Mean Absolute Percentage Error: 0.4276638457836126

 Polynomial Regression on all variables:
R-Squared: 0.49969296096402016
Mean Absolute Error: 46.30738007987664
Mean Absolute Percentage Error: 0.36207218444301953


## Conclusion

### Comparing R-Squared values
The R-squared identifies the proportion of the variance in the target variable with independent variables. The range of R-Squared value will be in between 0 and 1. Here, the R-Squared values are:
* Linear Regression = 0.606
* Polynomial Regression (BMI only) = 0.414
* Polynomial Regression (all variables) = 0.5
<br>Higher the R-Squared value refers to better performance on model in terms of target variable. In this case, the linear regression gives better results.

### Comparing Mean Absolute Percentage Error
MAPE is the average percentage difference with predicted and actual values, with respective of actual values. Lower the MAPE value shows better performance on model. 
* Linear Regression: 0.302
* Polynomial Regression (BMI only) = 0.428
* Polynomial Regression (all variables) = 0.362
<br> In the case linear regression gives better performance in terms os mean absolute percentage error.

### Comparing Mean Absolute Error
MAE  is the measurement of absolute average size of errors between predicted and actual values. Lower the MAE values indicates better model performance. 
8 Linear Regression: 40.295
* Polynomial Regression (BMI only) = 51.519
* Polynomial Regression (all variables) = 46.307
<br> In the case linear regression gives better performance in terms os mean absolute error.

##### Based on the considering all the comparison linear regression model better than other two, and provides a better model.

## Finding the Parameters of each type

In [28]:
linearParameters = X_train.shape[1] + 1
print("Number of Parameters by Linear Regression", linearParameters)
polyBMIParameters = len(polyBMI.get_feature_names_out())
print("Number of Parameters by Polynomial Regression on BMI only: ", polyBMIParameters)
polyAllParameters = len(polyAll.get_feature_names_out())
print("Number of Parameters by Polynomial Regression on all variables: ", polyAllParameters)

Number of Parameters by Linear Regression 11
Number of Parameters by Polynomial Regression on BMI only:  3
Number of Parameters by Polynomial Regression on all variables:  65


### Parameters of each type: <br>
* Linear Regression = 11
* Polynomial Regression (BMI only) = 3
* Polynomial Regression (all variables) = 65

<br> In multivariate linear regression, the data is fitted as a linear equation with multiple independent variables. The number of parameters will be equal to the number of features. For a polynomial regression on BMI only we have took 2 BMI feature alone, we will have 3 parameters. But while considering the polynomial regression on all variables, we define the polynomial features for all original features. So we will have 65 parameters here.

### My choice for deployment
<br> In  my choice for deployment based on these values, it will be multivariate linear regression. Since it shows better performance in all values as r-squared, mean absolute error, mean absolute percentage error values. Also with the number of parameters, linear regression has only limited parameters gives an advantage of less over fitting.